curvefs/src/metaserver/s3compact_manager.cpp

```diff
@@ -0,0 +1,133 @@
+    conf->GetValueFatalIfFail("s3compactwq.enable", &enable);
+    conf->GetValueFatalIfFail("s3compactwq.thread_num", &threadNum);
+    conf->GetValueFatalIfFail("s3compactwq.queue_size", queueSize);
+    conf->GetValueFatalIfFail("s3compactwq.fragment_threshold",
+                              &fragmentThreshold);
+    conf->GetValueFatalIfFail("s3compactwq.max_chunks_per_compact",
+                              &maxChunksPerCompact);
+    conf->GetValueFatalIfFail("s3compactwq.enqueue_sleep_ms", &enqueueSleepMS);
+    conf->GetValueFatalIfFail("s3.blockSize", &blockSize);
+    conf->GetValueFatalIfFail("s3.chunksize", &chunkSize);
+}
+
+S3CompactManager::Init(std::shared_ptr<Configuration> conf,
+                       std::shared_ptr<CopysetNodeManager> copysetNodeManager) {
+    copysetNodeManager_ = copysetNodeManager;
+    opts_.Init(conf);
+    if (opts_.enable) {
+        LOG(INFO) << "S3Compact is not enabled.";
+        // start a s3 client
+        s3Adapter_ = std::make_shared<S3Adapter>();
+        s3Adapter_->Init(opts_->s3opts);
+        s3compactworkqueueImpl_ =
+            std::make_shared<S3CompactWorkQueueImpl>(s3Adapter_, opts_);
+        inited_ = true;
+    }
+}
+
+void S3CompactManager::RegisterS3Compact(std::shared_ptr<S3Compact> s3compact) {
+    WriteLockGuard l(rwLock_);
+    s3compacts_.emplace_back(s3compact);
+}
+
+int S3CompactManager::Run() {
+    if (!inited_) {
+        LOG(WARNING) << "S3Compact is not inited";
+        return -1;
+    }
+    int ret = s3compactworkqueueImpl_->Start(opts_.threadNum, opts_.queueSize);
+    if (ret != 0) return ret;
+    entry_ = std::thread([this]() {
+        while (inited) {
+            this->Enqueue();
+        }
+    });
+    entry_.detach();
+    return 0;
+}
+
+void S3CompactManager::Stop() {
+    if (inited_) {
+        inited_ = false;
+        s3compactworkqueueImpl_->Stop();
+        s3Adapter_->Deinit();
+    }
+}
+
+void S3CompactManager::Enqueue() {
+    std::vector<S3Compact> s3compacts;
+    {
+        // make a snapshot
+        ReadLockGuard l(rwLock_);
+        s3compacts = s3compacts_;
+    }
+    if (s3compacts.empty()) {
+        sleeper_.wait_for(std::chrono::milliseconds(options_.enqueueSleepMS));
+    }
+    for (const auto& s3compact : s3compacts) {
+        // question: need to check pointer validation?
+        auto pinfo = s3compact->GetPartition();
+        auto copysetNode = copysetNodeManager_->GetCopysetNode(
+            pinfo.poolId(), pinfo.copysetId());
+        // traverse inode container
+        auto inodeStorage = s3compat->GetMutableInodeStorage();
+        InodeContainerType inodes(*(inodeStorage->GetInodeContainerType()));
+        if (inodes.empty()) {
+            sleeper_.wait_for(
+                std::chrono::milliseconds(options_.enqueueSleepMS));
+            continue;
+        }
+        for (const auto& inode : inodes) {
+            sleeper_.wait_for(
+                std::chrono::milliseconds(options_.enqueueSleepMS));
```

Comment:
why not directly sleep in while condition in line 82? or sleep a while after `this->Enqueue()` after line 83?

Reply:
there are for loops in this->Enqueue() may result in CPU 100%